text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+See the [FPC Shim](#fpc-shim) section for more information.
+
+**Ledger Enclave:** The Ledger Enclave is a crucial component to protect the execution of a FPC chaincode.
+In particular, it is responsible to ensure that a FPC chaincode processes committed world state data, i.e. only state data that comes from valid and committed transactions.
+The Ledger Enclave is a separate enclave that establishes a trusted view of the ledger and locally maintains integrity metadata for validating information on the blockchain ledger.
+Like the Peer, it performs standard validation logic when a new block of transactions arrives, but it also creates and stores a cryptographic hash of each new key-value pair of the blockchain state. This makes it possible to verify that the data coming from the Peer is correct -- remember that in the FPC setting we do not necessarily trust a (single) peer. The Chaincode Enclave uses the Ledger Enclave as its tamperproof source of information on the blockchain world state. The Ledger Enclave and a Chaincode Enclave interact by through a secure channel that is initially established when the Chaincode Enclave is deployed. There is one Ledger Enclave per Channel on any given Peer. When a peer receives a new block from the ordering service, it forwards it to the Ledger Enclave. The Ledger Enclave is part of the FPC runtime.
+
+**FPC Registry:** Also referred to as the Enclave Registry Chaincode (ERCC), this is a component which maintains a list of all Chaincode Enclaves deployed on the peers in a channel.
+The registry associates with each enclave their identity, associated public keys and an attestation linking them. Additionally, the registry manages chaincode specific keys, including a chaincode public encryption key, and facilitates corresponding key-management among authorized Chaincode Enclaves. Lastly, the registry also records information required to bootstrap the validation of attestation. All of this information is committed firmly on the ledger. This enables any Peer in the Channel (even those without SGX) to inspect the Attestation results before taking actions such as connecting to that chaincode or committing transactions produced by an FPC chaincode. Moreover, clients can query the registry to retrieve the chaincode public encryption keys of a particular FPC chaincode so they can send privately transaction proposals for endorsement.
+
+**FPC Validator:** This validation complements the Peer validation logic by validating transactions produced by an FPC Chaincode.
+It does this by checking that each transaction has been cryptographically signed (endorsed) by a Chaincode Enclave that is listed in the FPC Registry and is authorized for that chaincode.
```

Comment:
Does the validator know the public keys of the consensus service? Is it aware of what are Fabric block signatures, and all Fabric config changes? I am asking because if the validator has no ability to validate Fabric blocks on its own, then it's trivial to do a reorder attack (change order between transactions) and make it think something is invalid when it is valid.

Reply:
The validator exists at two places: as plugins in the peer and as part of the trusted-ledger. For plugins, they rely on the normal peer; for the validation logic inside the trusted ledger, they rely on the trusted ledger's core fabric validation logic which (a) explicitly references the (cryptographic) channel identity (and exposes this transitively to clients and chaincode enclaves) and (b) includes following/validating changes to ordering service info in channel definitions.